feat: Support Anthropic extended thinking and interleaved thinking

serialx opened the pull request:
Anthropic's API requires thinking blocks to be the first content in assistant messages when reasoning is enabled and tool calls are present.
ref: https://docs.anthropic.com/en/docs/build-with-claude/extended-thinking

This change:
- Stores thinking text in ResponseReasoningItem.content field
- Stores signature in encrypted_content field for metadata
- Reconstructs thinking blocks when converting items->messages
- Handles both streaming and non-streaming cases
- Only affects Anthropic models with reasoning enabled
- Adds regression test to prevent future breakage

Fixes compatibility with Claude models when using tools with interleaved thinking.

This fixes #765 
This fixes #678 

Code to reproduce:
```python
import asyncio

from openai.types.shared import Reasoning
from pydantic import BaseModel

from agents import (
    Agent,
    AgentUpdatedStreamEvent,
    ModelSettings,
    RawResponsesStreamEvent,
    RunItemStreamEvent,
    Runner,
    function_tool,
)


class WeatherTemp(BaseModel):
    city: str
    temperature_range: str


class WeatherConditions(BaseModel):
    city: str
    conditions: str


@function_tool
def get_weather_temp(city: str) -> WeatherTemp:
    """Get the current weather temperature information for a specified city."""
    print("[debug] get_weather_temp called")
    return WeatherTemp(city=city, temperature_range="14-20C")


@function_tool
def get_weather_conditions(city: str) -> WeatherConditions:
    """Get the current weather conditions information for a specified city."""
    print("[debug] get_weather_conditions called")
    return WeatherConditions(city=city, conditions="Sunny with wind.")


agent = Agent(
    name="Hello world",
    instructions="You are a helpful agent.",
    tools=[get_weather_temp, get_weather_conditions],
    model="litellm/anthropic/claude-sonnet-4-20250514",
    model_settings=ModelSettings(
        reasoning=Reasoning(effort="medium", summary="auto"),
        # Works with both with and without interleaved thinking
        extra_headers={"anthropic-beta": "interleaved-thinking-2025-05-14"},
    ),
)

PROMPT = (
    "What's the temperature in Tokyo? If the temp is below 30C, get the weather "
    "conditions and if it's sunny with wind, get the temperature in Tokyo's nearest "
    "neighbour country's capital"
)


async def agent_run():
    print("=== Runner.run ===")
    result = await Runner.run(
        agent,
        input=PROMPT,
    )
    print(result.final_output)
    print()


async def agent_run_streamed():
    print("=== Runner.run_streamed ===")
    result = Runner.run_streamed(
        agent,
        input=PROMPT,
    )
    async for event in result.stream_events():
        match event:
            case RawResponsesStreamEvent(data=data):
                delta = data.delta if hasattr(data, "delta") else ""
                print(f"Raw response: {data.type} {repr(delta)}")
            case RunItemStreamEvent():
                print("Run item:", event.type)
            case AgentUpdatedStreamEvent():
                print("Agent updated:", event.type)
    print(result.final_output)


async def main():
    await agent_run_streamed()
    await agent_run()


if __name__ == "__main__":
    asyncio.run(main())

```

Above code results in error in current main HEAD:
```
litellm.exceptions.BadRequestError: litellm.BadRequestError: AnthropicException - {"type":"error","error":{"type":"invalid_request_error","message":"messages.1.content.0.type: Expected `thinking` or `redacted_thinking`, but found `text`. When `thinking` is enabled, a final `assistant` message must start with a thinking block (preceeding the lastmost set of `tool_use` and `tool_result` blocks). We recommend you include thinking blocks from previous turns. To avoid this requirement, disable `thinking`. Please consult our documentation at https://docs.anthropic.com/en/docs/build-with-claude/extended-thinking"}}
```

serialx committed:
feat: Support Anthropic extended thinking and interleaved thinking

Anthropic's API requires thinking blocks to be the first content in assistant
messages when reasoning is enabled and tool calls are present.
ref: https://docs.anthropic.com/en/docs/build-with-claude/extended-thinking

This change:
- Stores thinking text in ResponseReasoningItem.content field
- Stores signature in encrypted_content field for metadata
- Reconstructs thinking blocks when converting items->messages
- Handles both streaming and non-streaming cases
- Only affects Anthropic models with reasoning enabled
- Adds regression test to prevent future breakage

Fixes compatibility with Claude models when using tools with interleaved thinking.

serialx commented on src/agents/models/chatcmpl_stream_handler.py:
```diff
@@ -527,7 +543,19 @@ async def handle_stream(
 
         # include Reasoning item if it exists
         if state.reasoning_content_index_and_output:
-            outputs.append(state.reasoning_content_index_and_output[1])
+            reasoning_item = state.reasoning_content_index_and_output[1]
+            # Store thinking text in content and signature in encrypted_content
+            if state.thinking_text:
+                # Add thinking text as a Content object
+                if not reasoning_item.content:
+                    reasoning_item.content = []
+                reasoning_item.content.append(
+                    Content(text=state.thinking_text, type="reasoning_text")
+                )
+            # Store signature in encrypted_content
+            if state.thinking_signature:
+                reasoning_item.encrypted_content = state.thinking_signature
```
@codex This is the same as above.

serialx committed:
refactor: more specific code path for thinking block insertion

Move the location of thinking block insertion from `ensure_assistant_message()` to `elif func_call := cls.maybe_function_tool_call(item):` to be more specific and prevent unintended side-effects.

serialx committed:
fix: handoff from reasoning model to non-reasoning

Be more selective/surgical on the reasoning message preservation.
When we handoff from Claude 4 Sonnet Thinking to non-thinking agent, we get errors because non-thinking models expects no thinking blocks in the request. This fixes this edge case by only preserving blocks when reasoning effort is not None.

seratch commented on src/agents/models/chatcmpl_converter.py:
```diff
@@ -483,9 +514,28 @@ def ensure_assistant_message() -> ChatCompletionAssistantMessageParam:
+                content_items = reasoning_item.get("content", [])
+                signature = reasoning_item.get("encrypted_content")
+
+                if content_items and preserve_reasoning_message:
```
this whole part is very much anthropic-specific logic, so can you rename the flag to `preserve_thinking_blocks`?

serialx committed:
refactor: rename preserve_reasoning_message to preserve_thinking_blocks

Since this parameter is anthropic-specific logic, changed the name to anthropic terminology

seratch reviewed (approved):
Confirmed this works with the model

Sourav-Nandy-ai commented:
Hi Sung-jin,

PR #1744 (plus the same-week follow-ups #1779 and #1784) is a good example of a change where correctness is easy to verify functionally, but subtle behavior drift is much harder to catch. If extended-thinking content is dropped or garbled across turns, it does not necessarily throw an error. It can simply quietly change which tool the agent decides to call next.

That is the class of problem agent-eval (agent_regress) is built to solve. Rather than relying on pass-or-fail metrics for single responses, it runs an agent N times before and after a change. It then compares the resulting behavior distributions statistically (using Mann-Whitney U, bootstrap CI, and effect size), returning a REGRESSED, STABLE, or IMPROVED verdict that can gate CI.

In the interest of not overselling, my OpenAI Agents SDK integration used to be non-functional. It called agent.run(), a method that does not exist on the real SDK, meaning it crashed before doing anything on every agent.

That is now fixed to call the real Runner.run() entrypoint, which is the same call your own PR's reproduction script uses. This resolves the crash that would have entirely blocked using this tool on a thinking-block change like yours.

I am genuinely curious and not just pitching a finished product. How did you validate that preserving and reconstructing thinking blocks did not shift tool-calling decisions across your test cases? If a distributional before-and-after comparison like this would have been useful, I would welcome your feedback on whether it is pointed at the right thing.

Best, 
Sourav
